coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala

```diff
@@ -0,0 +1,150 @@
+    val promQlParams = PromQlQueryParams(ConfigFactory.parseString(s"""endpoint = "${endPoint}""""),
+      queryParams.promQl, startMs / 1000, queryParams.stepSecs, endMs / 1000, queryParams.spread, processFailure = true,
+      processRouting = false)
+    PromQlExec(queryContext, InProcessPlanDispatcher, dataset.ref, promQlParams)
+  }
+
+  def materializeSimpleQuery(logicalPlan: LogicalPlan, qContext: QueryContext): ExecPlan = {
+
+    val routingKeys = getRoutingKeys(logicalPlan)
+    if (routingKeys.forall(_._2.isEmpty)) localPlanner.materialize(logicalPlan, qContext)
+    else {
+      val routingKeyMap = routingKeys.map(x => (x._1, x._2.get.head)).toMap
+      val offsetMs = LogicalPlanUtils.getOffsetMillis(logicalPlan)
+      val periodicSeriesTime = getPeriodicSeriesTimeFromLogicalPlan(logicalPlan)
+      val periodicSeriesTimeWithOffset = TimeRange(periodicSeriesTime.startMs - offsetMs,
+        periodicSeriesTime.endMs - offsetMs)
+      val lookBackMs = getLookBackMillis(logicalPlan)
+      val stepMs = logicalPlan.asInstanceOf[PeriodicSeriesPlan].stepMs
+      // Time at which raw data would be retrieved which is used to get partition assignments.
+      // It should have time with offset and lookback as we need raw data at time including offset and lookback.
+      val queryTimeRange = TimeRange(periodicSeriesTimeWithOffset.startMs - lookBackMs,
+        periodicSeriesTimeWithOffset.endMs)
+
+      val partitions = partitionLocationProvider.getPartitions(routingKeyMap, queryTimeRange).
+        sortBy(_.timeRange.startMs)
+      var prevPartitionStart = periodicSeriesTimeWithOffset.startMs
+      val execPlans = partitions.zipWithIndex.map { case (p, i) =>
+        // First partition should start from query start time
+        val startMs = if (i == 0) periodicSeriesTime.startMs
+                      else {
+                        val numStepsInPrevPartition = (p.timeRange.startMs - prevPartitionStart + lookBackMs) / stepMs
+                        val lastPartitionInstant = prevPartitionStart + numStepsInPrevPartition * stepMs
+                        lastPartitionInstant + stepMs
+                      }
+        prevPartitionStart = startMs
+        val endMs = p.timeRange.endMs + offsetMs
+        if (p.partitionName.equals(localPartition)) localPlanner.materialize(
+          copyWithUpdatedTimeRange(logicalPlan, TimeRange(startMs, endMs)), qContext)
+        else generateRemoteExec(qContext, startMs, endMs, p.endPoint)
+      }
+      if (execPlans.size == 1) execPlans.head
+      else StitchRvsExec(qContext, InProcessPlanDispatcher, execPlans.sortWith((x, y) => !x.isInstanceOf[PromQlExec]))
+      // ^^ Stitch RemoteExec plan results with local using InProcessPlanDispatcher
+      // Sort to move RemoteExec in end as it does not have schema
+    }
+  }
+
+
+  def materializeBinaryJoin(logicalPlan: LogicalPlan, qContext: QueryContext): ExecPlan = {
+
+    val routingKeys = getRoutingKeys(logicalPlan)
```

Comment:
Can you move all the common logic between `materializeBinaryJoin` and `materializeSimpleQuery` to helper methods? Looks like line#98 to line#114 is similar across both the methods.